src/cpuid/src/amd/mod.rs

```diff
@@ -0,0 +1,301 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+#![warn(clippy::pedantic)]
+#![allow(
+    clippy::similar_names,
+    clippy::unreadable_literal,
+    clippy::module_name_repetitions
+)]
+
+#[cfg(cpuid)]
+use super::{CCpuidResult, KvmCpuidFlags};
+use super::{CpuidEntry, CpuidKey, CpuidTrait, RawCpuid, RawKvmCpuidEntry, Supports};
+
+/// Error types.
+mod errors;
+pub use errors::*;
+
+/// Register bit fields.
+mod registers;
+pub use registers::*;
+
+/// Leaf structs.
+mod leaves;
+pub use leaves::*;
+
+/// Indexing implementations.
+mod indexing;
+pub use indexing::*;
+
+/// A structure containing the information as described in the AMD CPUID specification as described
+/// in
+/// [AMD64 Architecture Programmer’s Manual Volume 3: General-Purpose and System Instructions](https://www.amd.com/system/files/TechDocs/24594.pdf)
+/// .
+///
+/// # Notes
+///
+/// We not do not currently check AMD features on snapshot restore.
+#[derive(Debug, Clone, Eq, PartialEq, construct::Inline)]
+#[repr(C)]
+pub struct AmdCpuid(pub std::collections::BTreeMap<CpuidKey, CpuidEntry>);
+
+impl CpuidTrait for AmdCpuid {
+    /// Gets a given sub-leaf.
+    #[inline]
+    fn get(&self, key: &CpuidKey) -> Option<&CpuidEntry> {
+        self.0.get(key)
+    }
+    /// Gets a given sub-leaf.
+    #[inline]
+    fn get_mut(&mut self, key: &CpuidKey) -> Option<&mut CpuidEntry> {
+        self.0.get_mut(key)
+    }
+}
+
+impl AmdCpuid {
+    /// We always use this brand string.
+    pub const DEFUALT_BRAND_STRING: &[u8; super::BRAND_STRING_LENGTH] =
+        b"AMD EPYC\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0";
+
+    /// Applies `vm_spec` to `self`.
+    ///
+    /// # Errors
+    ///
+    /// When attempting to access misisng leaves or set fields within leaves to values that don't
+    /// fit.
+    ///
+    /// # Panics
+    ///
+    /// Never.
+    // As we pass through host frequency, we require CPUID and thus `cfg(cpuid)`.
+    #[allow(clippy::too_many_lines)]
+    #[cfg(cpuid)]
+    #[inline]
+    pub fn normalize(
```

Comment:
couldn't `normalize` be a member of the `CpuidTrait`, since we seem to implement it in every case?

Reply:
`normalize` serves the same purpose in `IntelCpuid` and `AmdCpuid` but it doesn't share any functionality. `CpuidTrait` is implemented to reduce code duplication by sharing functionality. Moving `normalize` into this trait wouldn't reduce the amount of code, but as with using traits in general it would make it slightly more complex, so I don't think it makes sense in my opinion.